serial.c

```diff
@@ -452,6 +490,235 @@ static void Interrupt com_general_isr(void)
 /* ======================================================================== */
 
 
+#ifdef PROTECTED_MODE
+
+/* we do not know the exact size of com_general_isr() function
+   but for sure it's not longer then 2048 bytes */
+#define ISR_SIZE        2048
+
+#if defined( __DJGPP__ )
+
+static void serial_dpmi_get_pvect(int vector, _go32_dpmi_seginfo *info)
+{
+    _go32_dpmi_get_protected_mode_interrupt_vector(vector, info);
+}
+
+static void serial_dpmi_set_pvect(int vector, _go32_dpmi_seginfo *info)
+{
+    _go32_dpmi_set_protected_mode_interrupt_vector(vector, info);
+}
+
+static int serial_dpmi_lock_memory(void)
+{
+    unsigned long dataaddr, codeaddr;
+    __dpmi_meminfo dataregion, coderegion;
+
+    if(__dpmi_get_segment_base_address(_my_cs(), &codeaddr) == 0 &&
+       __dpmi_get_segment_base_address(_my_ds(), &dataaddr) == 0)
+    {
+        coderegion.handle = 0;
+        coderegion.size = ISR_SIZE;
+        coderegion.address = codeaddr + (unsigned long)com_general_isr;
+        dataregion.handle = 0;
+        dataregion.size = sizeof(g_comports);
+        dataregion.address = codeaddr + (unsigned long)g_comports;
+        if(__dpmi_lock_linear_region(&coderegion) == 0)
+        {
+            if(__dpmi_lock_linear_region(&dataregion) == 0)
+            {
+                g_isr_addr.pm_offset = (unsigned long)com_general_isr;
+                g_isr_addr.pm_selector = _go32_my_cs();
+                if(_go32_dpmi_allocate_iret_wrapper(&g_isr_addr) == 0)
+                    return 1;
+                __dpmi_unlock_linear_region(&dataregion);
+            }
+            __dpmi_unlock_linear_region(&coderegion);
+        }
+    }
+    return 0;
+}
+
+static void serial_dpmi_unlock_memory(void)
+{
+    unsigned long baseaddr;
+    __dpmi_meminfo region;
+
+    if(__dpmi_get_segment_base_address(_my_ds(), &baseaddr) == 0)
+    {
+        region.handle = 0;
+        region.size = sizeof(g_comports);
+        region.address = baseaddr + (unsigned long)g_comports;
+        __dpmi_unlock_linear_region(&region);
+    }
+    if(__dpmi_get_segment_base_address(_my_cs(), &baseaddr) == 0)
+    {
+        region.handle = 0;
+        region.size = ISR_SIZE;
+        region.address = baseaddr + (unsigned long)com_general_isr;
+        __dpmi_unlock_linear_region(&region);

```

Comment:
I think that any time we call an API that can return failure, we should pass notification of a failure up to the caller unless we ourselves have a way to fix the problem. Then we have done due diligence and it's up to the caller to decide what to do about it (if anything at all).

I wouldn't worry too much about a dummy return value. It just means that the real mode version cannot fail, whereas the protected mode version can.